src/Stratis.SmartContracts.Executor.Reflection/ILRewrite/ObserverRewriter.cs

```diff
@@ -0,0 +1,272 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Mono.Cecil;
+using Mono.Cecil.Cil;
+using RuntimeObserver;
+
+namespace Stratis.SmartContracts.Executor.Reflection.ILRewrite
+{
+    /// <summary>
+    /// Rewrites a module to include an <see cref="Observer"/> that tracks gas usage. 
+    /// </summary>
+    public class ObserverRewriter : IILRewriter
+    {
+        private static readonly HashSet<OpCode> BranchingOps = new HashSet<OpCode>
+        {
+            OpCodes.Beq,
+            OpCodes.Beq_S,
+            OpCodes.Bge,
+            OpCodes.Bge_S,
+            OpCodes.Bge_Un,
+            OpCodes.Bge_Un_S,
+            OpCodes.Bgt,
+            OpCodes.Bgt_S,
+            OpCodes.Ble,
+            OpCodes.Ble_S,
+            OpCodes.Ble_Un,
+            OpCodes.Blt,
+            OpCodes.Bne_Un,
+            OpCodes.Bne_Un_S,
+            OpCodes.Br,
+            OpCodes.Brfalse,
+            OpCodes.Brfalse_S,
+            OpCodes.Brtrue,
+            OpCodes.Brtrue_S,
+            OpCodes.Br_S
+        };
+
+        private static readonly HashSet<OpCode> CallingOps = new HashSet<OpCode>
+        {
+            OpCodes.Call,
+            OpCodes.Calli,
+            OpCodes.Callvirt
+        };
+
+        /// <summary>
+        /// Holds the key to retrieve the last Observer instance created from <see cref="ObserverInstances"/>. 
+        /// 
+        /// TODO: There should be a better pattern here to implement the Rewriter interface and still be able to retrieve this.
```

Comment:
We could remove `lastRewritten` by injecting the Observer instance with the gasMeter attached. if we're not DI-ing the re-writer, we can just pass it into the constructor. If we are we will need an ObserverRewriterFactory.

Reply:
This will also enable us to call `ObserverInstances.Set` inside the rewriter.